src/abe/dippe.c

```diff
@@ -0,0 +1,613 @@
+    cfe_uniform_sample_vec(&A, dippe->p);
+
+    // g1^A (k+1 x k)
+    dippe->g1_A = (ECP_BN254*)cfe_malloc(sizeof(ECP_BN254) * (assump_size+1) * assump_size);
+    for (unsigned int i=0; i<(assump_size+1); i++) {
+        for (unsigned int j=0; j<assump_size; j++) {
+            ECP_BN254_generator(&(dippe->g1_A[(assump_size*i)+j]));
+            BIG_256_56_from_mpz(tmp_big, A.vec[(assump_size*i)+j]);
+            ECP_BN254_mul(&(dippe->g1_A[(assump_size*i)+j]), tmp_big);
+        }
+    }
+
+    // U (k+1 x k+1)
+    cfe_vec U;
+    cfe_vec_init(&U, ((assump_size+1) * (assump_size+1)));
+    cfe_uniform_sample_vec(&U, dippe->p);
+
+    // U(T)A  (k+1 x k)
+    cfe_vec UA;
+    cfe_vec_init(&UA, ((assump_size+1) * assump_size));
+    for (unsigned int i=0; i<(assump_size+1); i++) {
+        for (unsigned int j=0; j<assump_size; j++) {
+            mpz_set_ui(UA.vec[(assump_size*i)+j], 0);
+            for (unsigned int k=0; k<(assump_size+1); k++) {
+                mpz_mul(tmp_mpz, U.vec[i+(k*(assump_size+1))], A.vec[(assump_size*k)+j]);
+                mpz_add(UA.vec[(assump_size*i)+j], UA.vec[(assump_size*i)+j], tmp_mpz);
+                mpz_mod(UA.vec[(assump_size*i)+j], UA.vec[(assump_size*i)+j], dippe->p);
+            }
+        }
+    }
+
+    // g1^U(T)A (k+1 x k)
+    dippe->g1_UA = (ECP_BN254*)cfe_malloc(sizeof(ECP_BN254) * (assump_size+1) * assump_size);
+    for (unsigned int i=0; i<(assump_size+1); i++) {
+        for (unsigned int j=0; j<assump_size; j++) {
+            ECP_BN254_generator(&(dippe->g1_UA[(assump_size*i)+j]));
+            BIG_256_56_from_mpz(tmp_big, UA.vec[(assump_size*i)+j]);
+            ECP_BN254_mul(&(dippe->g1_UA[(assump_size*i)+j]), tmp_big);
+        }
+    }
+
+    // cleanup
+    cfe_vec_free(&A);
+    cfe_vec_free(&U);
+    cfe_vec_free(&UA);
+    mpz_clear(tmp_mpz);
+}
+
+/**
+ * Initializes an authority and generates its public and private key
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param pk A pointer to a cfe_dippe_pub_key struct; Represents the public key that is about to be populated
+ * @param sk A pointer to a cfe_dippe_sec_key struct; Represents the secret key that is about to be populated
+ */
+void cfe_dippe_authsetup(cfe_dippe *dippe, cfe_dippe_pub_key *pk, cfe_dippe_sec_key *sk) {
+
+    // temp vars
+    ECP_BN254  tmp_g1;
+    FP12_BN254 tmp_gt;
+    BIG_256_56 tmp_big;
+    ECP2_BN254 tmp_g2;
+
+    ECP2_BN254_generator(&tmp_g2);
+
+    ///////////////////////////////////////////////
+    // Secret key section
+    ///////////////////////////////////////////////
+
+    // sigma
+    mpz_init(sk->sigma);
+    cfe_uniform_sample(sk->sigma, dippe->p);
+
+    // alpha (k+1 x 1)
+    cfe_vec_init(&(sk->alpha), (dippe->assump_size+1));
+    cfe_uniform_sample_vec(&(sk->alpha), dippe->p);
+
+    // W (k+1 x k+1)
+    cfe_vec_init(&(sk->W), (dippe->assump_size+1) * (dippe->assump_size+1));
+    cfe_uniform_sample_vec(&(sk->W), dippe->p);
```

Comment:
Same comment here as above, you can use matrices instead of vectors.

Reply:
Done.